beacon_node/eth2_libp2p/src/discovery/mod.rs

```diff
@@ -519,69 +549,116 @@ impl<TSpec: EthSpec> Discovery<TSpec> {
 
     /// Drives the queries returning any results from completed queries.
-    fn poll_queries(&mut self, cx: &mut Context) -> Option<(Option<Instant>, Vec<Enr>)> {
+    fn poll_queries(&mut self, cx: &mut Context) -> Option<HashMap<PeerId, Option<Instant>>> {
+        let mut results: Vec<(Option<Instant>, Vec<Enr>)> = Vec::new();
+
         while let Poll::Ready(Some(query_future)) = self.active_queries.poll_next_unpin(cx) {
-            match query_future.0 {
-                QueryType::FindPeers => {
-                    self.find_peer_active = false;
-                    match query_future.1 {
-                        Ok(r) if r.is_empty() => {
-                            debug!(self.log, "Discovery query yielded no results.");
-                        }
-                        Ok(r) => {
-                            debug!(self.log, "Discovery query completed"; "peers_found" => r.len());
-                            return Some((None, r));
+            // Iterate through queries comprising the original grouped query
+            query_future.0.clone().into_iter().for_each(|query| {
+                let query_future_enrs = query_future.1.clone();
+                match query {
+                    QueryType::FindPeers => {
+                        self.find_peer_active = false;
+                        match query_future_enrs {
+                            Ok(r) if r.is_empty() => {
+                                debug!(self.log, "Discovery query yielded no results.");
+                            }
+                            Ok(r) => {
+                                debug!(self.log, "Discovery query completed"; "peers_found" => r.len());
+                                results.push((None, r));
+                            }
+                            Err(e) => {
+                                warn!(self.log, "Discovery query failed"; "error" => e.to_string());
+                            }
                         }
-                        Err(e) => {
-                            warn!(self.log, "Discovery query failed"; "error" => e.to_string());
+                    }
+                    QueryType::Subnet(subnet_query) => {
+                        match query_future_enrs {
+                            Ok(r) if r.is_empty() => {
+                                debug!(self.log, "Subnet discovery query yielded no results."; "subnet_id" => *subnet_query.subnet_id, "retries" => subnet_query.retries);
+                            }
+                            Ok(r) => {
+                                let enr: Vec<Enr> = r.clone();
+                                debug!(self.log, "Peer subnet discovery request completed"; "peers_found" => r.len(), "subnet_id" => *subnet_query.subnet_id);
+                                // A subnet query has completed. Add back to the queue, incrementing retries.
+                                self.add_subnet_query(subnet_query.subnet_id, subnet_query.min_ttl, subnet_query.retries + 1);
```

Comment:
This adds the individual subnet query back and increments the retries. They later get re-grouped when taken from the queue right?

This makes sense to me, they could be grouped in different ways and are independently failed.

Reply:
Yep. One thing I wasn't sure about is whether we want to check the retry count and whether we've reached the target peer count for a subnet before we add queries to the queue or after we pop queries off the queue.

Reply:
Yeah. I have it so that once we pop them off, but I did it from memory to not duplicate code. Because we need to check when we first add them and we need to check once their done. 

If you can find a clean way of doing it (with minimal code duplication) go for it